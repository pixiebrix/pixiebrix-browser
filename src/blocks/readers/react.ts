/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import { createSendScriptMessage } from "@/messaging/chrome";
import { READ_REACT_COMPONENT } from "@/messaging/constants";
import { ReaderOutput } from "@/core";
import { registerFactory } from "@/blocks/readers/factory";

export interface ReactConfig {
  type: "react";
  selector: string;
  traverseUp?: number;
  rootProp?: string;
  waitMillis?: number;
}

export const withReactComponent = createSendScriptMessage<ReaderOutput>(
  READ_REACT_COMPONENT
);

async function doRead(reader: ReactConfig): Promise<ReaderOutput> {
  const { selector, traverseUp = 0, waitMillis = 1000, rootProp } = reader;
  return await withReactComponent({
    selector,
    traverseUp,
    waitMillis,
    rootProp,
  });
}

registerFactory("react", doRead);
