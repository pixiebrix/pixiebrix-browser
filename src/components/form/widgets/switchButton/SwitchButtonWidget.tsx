/*
 * Copyright (C) 2023 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import React, { type ChangeEvent } from "react";
import BootstrapSwitchButton from "bootstrap-switch-button-react";
import { type CustomFieldWidget } from "@/components/form/FieldTemplate";
import styles from "./SwitchButtonWidget.module.scss";

export type CheckBoxLike = {
  name: string;
  value: boolean;
};

const SwitchButtonWidget: CustomFieldWidget<boolean, CheckBoxLike> = ({
  name,
  onChange,
  value,
  disabled,
}) => {
  const patchedOnChange = (checked: boolean) => {
    onChange({
      target: { value: checked, name },
    } as ChangeEvent<CheckBoxLike>);
  };

  return (
    <div className={styles.root}>
      <BootstrapSwitchButton
        disabled={disabled}
        onlabel=" "
        offlabel=" "
        checked={value}
        onChange={patchedOnChange}
      />
    </div>
  );
};

export default SwitchButtonWidget;
