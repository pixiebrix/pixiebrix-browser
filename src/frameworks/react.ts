/*
 * Copyright (C) 2020 Pixie Brix, LLC
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import fromPairs from "lodash/fromPairs";

interface DOMFiber {
  _instance: any;
  _currentElement: {
    _owner: DOMFiber;
  };
}

interface ComponentFiber {
  return: ComponentFiber;
  memoizedProps: { [key: string]: unknown };
}

export class ComponentNotFoundError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ComponentNotFoundError";
  }
}

export function readReactProps(
  fiber: ComponentFiber,
  rootProp?: string
): Record<string, unknown> {
  // getComponentFiber is already returning the "return" for the component
  const memoizedProps = fiber.memoizedProps;
  console.debug("Memoized props", memoizedProps);
  const props = rootProp ? memoizedProps[rootProp] : memoizedProps;
  return fromPairs(Object.entries(props).filter(([key]) => key !== "children"));
}

export function findReactComponent(
  dom: HTMLElement,
  traverseUp = 0
): ComponentFiber {
  // https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging
  // https://stackoverflow.com/a/39165137/402560

  const key = Object.keys(dom).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );
  // @ts-ignore: works in practice
  const domFiber: DOMFiber | null = dom[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found at selector");
  }

  // react <16
  // TODO: find a react 15.0 app to test this on
  if (domFiber._currentElement) {
    throw new Error("Support for React <16 not implemented");
    // let componentFiber = domFiber._currentElement._owner;
    // for (let i = 0; i < traverseUp; i++) {
    //     componentFiber = componentFiber._currentElement._owner;
    // }
    // return componentFiber._instance;
  }

  // react 16+
  const getComponentFiber = (fiber: any) => {
    //return fiber._debugOwner; // this also works, but is __DEV__ only
    let parentFiber = fiber.return;
    while (typeof parentFiber.type == "string") {
      parentFiber = parentFiber.return;
    }
    return parentFiber;
  };
  let componentFiber = getComponentFiber(domFiber);
  for (let i = 0; i < traverseUp; i++) {
    componentFiber = getComponentFiber(componentFiber);
  }
  return componentFiber;
}
