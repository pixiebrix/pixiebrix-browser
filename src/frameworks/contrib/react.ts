/*
 * Copyright (C) 2021 PixieBrix, Inc.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Affero General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Affero General Public License for more details.
 *
 * You should have received a copy of the GNU Affero General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

import { pickBy } from "lodash";
import { ComponentNotFoundError, ignoreNotFound } from "@/frameworks/errors";
import { RootInstanceVisitor } from "@/frameworks/scanner";
import { ReadableComponentAdapter, traverse } from "@/frameworks/component";
import { isNode } from "@/frameworks/dom";

// React architecture references:
// https://github.com/acdlite/react-fiber-architecture
// https://indepth.dev/posts/1008/inside-fiber-in-depth-overview-of-the-new-reconciliation-algorithm-in-react#memoizedprops
// https://github.com/Venryx/mobx-devtools-advanced/blob/master/Docs/TreeTraversal.md
// https://stackoverflow.com/questions/29321742/react-getting-a-component-from-a-dom-element-for-debugging

interface RootInstance {
  _reactRootContainer: unknown;
}

/**
 * Structure from:
 * - https://github.com/facebook/react/blob/50d9451f320a9aaf94304209193562cc385567d8/packages/react-devtools-shared/src/backend/legacy/renderer.js#L739
 * - https://github.com/facebook/react/blob/v15.0.0/src/renderers/shared/reconciler/ReactReconciler.js
 */
interface LegacyInstance {
  /**
   * Public instance
   */
  _instance: unknown;

  _currentElement: {
    _owner: LegacyInstance;
  };
}

/**
 * Type definition from: https://github.com/facebook/react/blob/4c6470cb3b821f3664955290cd4c4c7ac0de733a/packages/react-reconciler/src/ReactInternalTypes.js#L62
 */
interface Fiber {
  /**
   * The Fiber to return to after finishing processing this one.
   * This is effectively the parent, but there can be multiple parents (two)
   * so this is only the parent of the thing we're currently processing.
   * It is conceptually the same as the return address of a stack frame.
   */
  return: Fiber | null;

  /**
   * The props used to create the output.
   */
  memoizedProps: { [prop: string]: unknown };

  /**
   * The resolved function/class associated with this fiber.
   */
  type: string | object;

  /**
   * The local state associated with this fiber.
   */
  stateNode: Node | object;
}

export function isManaged(node: Node): boolean {
  return Object.keys(node).some((key) =>
    key.startsWith("__reactInternalInstance$")
  );
}

export function hasReactProps(fiber: Fiber): boolean {
  return Object.keys(fiber.memoizedProps).some((x) => x !== "children");
}

export function readReactProps(fiber: Fiber): { [prop: string]: unknown } {
  return pickBy(fiber.memoizedProps, (value, key) => key !== "children");
}

function getComponentFiber(fiber: Fiber): Fiber {
  // return fiber._debugOwner; // this also works, but is __DEV__ only
  let parentFiber = fiber.return;
  while (typeof parentFiber.type === "string") {
    // string for HTML nodes, so traverse
    parentFiber = parentFiber.return;
  }
  return parentFiber;
}

export function findReactComponent(node: Node, traverseUp = 0): Fiber {
  // https://stackoverflow.com/a/39165137/402560
  const key = Object.keys(node).find((key) =>
    key.startsWith("__reactInternalInstance$")
  );

  const domFiber: Fiber | LegacyInstance | null = (node as any)[key];

  if (domFiber == null) {
    throw new ComponentNotFoundError("React fiber not found for element");
  }

  if ("_currentElement" in domFiber) {
    // FIXME: test this for React <16
    const owner = (x: LegacyInstance) => x._currentElement._owner;
    const fiber = traverse(owner, owner(domFiber), traverseUp);
    return fiber._instance as Fiber;
  } else {
    return traverse(getComponentFiber, getComponentFiber(domFiber), traverseUp);
  }
}

export class ReactRootVisitor implements RootInstanceVisitor<RootInstance> {
  public rootInstances: RootInstance[] = [];
  visit(node: Element | Node): boolean {
    if ("_reactRootContainer" in node) {
      this.rootInstances.push(node);
      return false;
    }
    return true;
  }
}

export const adapter: ReadableComponentAdapter<Fiber> = {
  isManaged,
  getComponent: (node) => ignoreNotFound(() => findReactComponent(node, 0)),
  getParent: getComponentFiber,
  getNode: (instance) =>
    isNode(instance.stateNode) ? instance.stateNode : null,
  getData: readReactProps,
  hasData: hasReactProps,
};

export default adapter;
