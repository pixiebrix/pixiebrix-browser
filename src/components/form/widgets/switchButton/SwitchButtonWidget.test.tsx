import { fireEvent, render } from "@testing-library/react";
import React from "react";
import SwitchButtonWidget from "./SwitchButtonWidget";

const values = [true, false];
test.each(values)("renders value %s", (value) => {
  const rendered = render(
    <SwitchButtonWidget value={value} onChange={jest.fn()} />
  );
  expect(rendered.asFragment()).toMatchSnapshot();
});

test.each(values)("calls onChange", (value) => {
  const name = "Name for Test";
  const onChangeMock = jest.fn();
  const { container } = render(
    <SwitchButtonWidget value={value} onChange={onChangeMock} name={name} />
  );

  fireEvent.click(container.querySelector(".switch"));

  expect(onChangeMock).toHaveBeenCalledWith({
    target: {
      value: !value,
      name,
    },
  });
});
