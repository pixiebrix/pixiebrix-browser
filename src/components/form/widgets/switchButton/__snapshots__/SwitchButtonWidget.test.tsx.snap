// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders value false 1`] = `
<DocumentFragment>
  <div
    class="switch btn off btn-light"
  >
    <div
      class="switch-group"
    >
      <span
        class="switch-on btn btn-primary"
      >
         
      </span>
      <span
        class="switch-off btn btn-light"
      >
         
      </span>
      <span
        class="switch-handle btn btn-light"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders value true 1`] = `
<DocumentFragment>
  <div
    class="switch btn on btn-primary"
  >
    <div
      class="switch-group"
    >
      <span
        class="switch-on btn btn-primary"
      >
         
      </span>
      <span
        class="switch-off btn btn-light"
      >
         
      </span>
      <span
        class="switch-handle btn btn-light"
      />
    </div>
  </div>
</DocumentFragment>
`;
